lib/Sema/CSSimplify.cpp

```diff
@@ -1543,7 +1546,8 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
         // Simplify the right-hand type and perform the "occurs" check.
         typeVar1 = getRepresentative(typeVar1);
         type2 = simplifyType(type2, flags);
-        if (typeVarOccursInType(typeVar1, type2))
+        if (typeVarOccursInType(typeVar1, type2) ||
+            type2->is<DependentMemberType>())
```

Comment:
Absolutely. I'll do that as an NFC follow-up commit, because I don't want to disrupt the run of the compatibility suite.